core/src/test/java/brooklyn/test/entity/RecordingSensorEventListener.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package brooklyn.test.entity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+import brooklyn.entity.Entity;
+import brooklyn.event.Sensor;
+import brooklyn.event.SensorEvent;
+import brooklyn.event.SensorEventListener;
+
+public class RecordingSensorEventListener<T> implements SensorEventListener<T> {
+    
+    private final boolean skipDuplicateValues;
+    private final List<SensorEvent<T>> events = Collections.synchronizedList(Lists.<SensorEvent<T>>newArrayList());
+    private final List<T> values = Collections.synchronizedList(Lists.<T>newArrayList());
+    private final Multimap<EntityAndSensor<?>, SensorEvent<T>> eventsBySource = 
+            Multimaps.synchronizedListMultimap(ArrayListMultimap.<EntityAndSensor<?>, SensorEvent<T>>create());
+    private final Multimap<EntityAndSensor<?>, T> valuesBySource = 
+            Multimaps.synchronizedListMultimap(ArrayListMultimap.<EntityAndSensor<?>, T>create());
+
+    public RecordingSensorEventListener() {
+        this(false);
+    }
+    
+    public RecordingSensorEventListener(boolean skipDuplicateValues) {
+        this.skipDuplicateValues = skipDuplicateValues;
+    }
+    
+    @Override
+    public void onEvent(SensorEvent<T> event) {
+        EntityAndSensor<?> source = new EntityAndSensor<T>(event.getSource(),  event.getSensor());
+        Collection<T> previousValues = valuesBySource.get(source);
+        if (skipDuplicateValues && !previousValues.isEmpty() && Objects.equal(Iterables.get(previousValues, previousValues.size()-1), event.getValue())) {
```

Comment:
Also also: worth clarifying that "duplicates" really means "consecutive duplicates".